server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java

```diff
@@ -0,0 +1,1048 @@
+            if (secStorageVm == null) {
+                LOGGER.warn("secondary storage VM " + cssHost.getName() + " doesn't exist");
+                return null;
+            }
+        }
+        DataStore ssStore = _dataStoreMgr.getImageStore(zone);
+        return ssStore.getUri();
+    }
+
+    public Map<String, Object> getConfigParams() {
+        return configParams;
+    }
+
+    public void setConfigParams(Map<String, Object> configParams) {
+        this.configParams = configParams;
+    }
+
+    @Override
+    public String getConfigComponentName() {
+        return RetrieveDiagnosticsServiceImpl.class.getSimpleName();
+    }
+
+    public Map<String, List<DiagnosticsKey>> getDefaultDiagnosticsData() {
+        return allDefaultDiagnosticsTypeKeys;
+    }
+
+    @Override
+    public ConfigKey<?>[] getConfigKeys()
+    {
+        return new ConfigKey<?>[] { RetrieveDiagnosticsTimeOut };
+    }
+
+
+    @Override
+    public List<Class<?>> getCommands(){
+        List<Class<?>> cmdList = new ArrayList<Class<?>>();
+        cmdList.add(RetrieveDiagnosticsCmd.class);
+        return cmdList;
+    }
+
+    protected class DiagnosticsGarbageCollector implements Runnable {
+        private String zipFileName = null;
+        private String path = "/diagnosticsdata/";
+        private Answer answer = null;
+
+
+        public DiagnosticsGarbageCollector() {
+        }
+
+        @Override
+        public void run() {
+            try {
+                LOGGER.trace("Diagnostics Garbage Collection Thread is running.");
+                File zipFile = new File(path + zipFileName);
+                Path zipFilePath = zipFile.toPath();
+                BasicFileAttributes attributes = null;
+                try {
+                    attributes = Files.readAttributes(zipFilePath, BasicFileAttributes.class);
+                } catch (IOException exception) {
+                    LOGGER.error("Could not find the zip file.", exception);
+                }
+                long milliseconds = attributes.creationTime().to(TimeUnit.MILLISECONDS);
+                VolumeVO volume = null;
+                volume = _volumeDao.findById(hostId);
+                Long zoneId = volume.getDataCenterId();
+                DataStore store = _dataStoreMgr.getImageStore(zoneId);
+                if (store == null) {
+                    throw new CloudRuntimeException("cannot find an image store for zone " + zoneId);
+                }
+                DataStoreTO destStoreTO = store.getTO();
+                if (milliseconds > RetrieveDiagnosticsFileAge.value()) {
+                    DeleteZipCommand cmd = new DeleteZipCommand(zipFilePath.toString(), destStoreTO);
+                    answer = cleanupDiagnostics(cmd);
+                }
+
+            } catch (Exception e) {
+                LOGGER.error("Caught the following Exception", e);
+            }
+        }
+    }
+
+    @Override
+    public Answer cleanupDiagnostics(DeleteZipCommand cmd) {
```

Comment:
Is this supposed to run in the server or in the agent?
if in the agent it shouldn't be implemented here.
if on the management server there is no need for the Command/Answer pattern
Can you explain your thinking here?